File: app/src/main/java/com/chooloo/www/callmanager/ui/settings_fragment/SettingsFragment.java
package com.chooloo.www.callmanager.ui.settings_fragment;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.telephony.SubscriptionInfo;
import android.telephony.SubscriptionManager;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.preference.CheckBoxPreference;
import androidx.preference.ListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.SwitchPreference;

import com.chooloo.www.callmanager.R;
import com.chooloo.www.callmanager.ui.main.MainActivity;

import java.util.ArrayList;
import java.util.List;

import timber.log.Timber;

import static android.Manifest.permission.READ_PHONE_STATE;

public class SettingsFragment extends PreferenceFragmentCompat implements SettingsMvpView {

    private SettingsPresenter<SettingsMvpView> mPresenter;

    public static SettingsFragment newInstance() {
        Bundle args = new Bundle();
        SettingsFragment fragment = new SettingsFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        onSetup();
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.preference, rootKey);

        findPreference(getString(R.string.pref_app_color_key)).setOnPreferenceChangeListener((preference, newValue) -> {
            mPresenter.refresh();
            return mPresenter.onListPreferenceChange(preference, newValue);
        });

        findPreference(getString(R.string.pref_app_theme_key)).setOnPreferenceChangeListener((preference, newValue) -> {
            mPresenter.refresh();
            return mPresenter.onListPreferenceChange(preference, newValue);
        });

        findPreference(getString(R.string.pref_reject_call_timer_key)).setOnPreferenceChangeListener((preference, newValue) -> mPresenter.onListPreferenceChange(preference, newValue));
        findPreference(getString(R.string.pref_answer_call_timer_key)).setOnPreferenceChangeListener(((preference, newValue) -> mPresenter.onListPreferenceChange(preference, newValue)));
        findPreference(getString(R.string.pref_default_page_key)).setOnPreferenceChangeListener((preference, newValue) -> mPresenter.onListPreferenceChange(preference, newValue));
        findPreference(getString(R.string.pref_sim_select_key)).setOnPreferenceChangeListener((preference, newValue) -> mPresenter.onListPreferenceChange(preference, newValue));
        findPreference(getString(R.string.pref_is_biometric_key)).setOnPreferenceChangeListener((preference, newValue) -> mPresenter.onSwitchPreferenceChange(preference, newValue));
        setupSimSelection();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        mPresenter.onRequestPermissionResult(requestCode, grantResults);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mPresenter.attach(this);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mPresenter.detach();
    }

    @Override
    public void onSetup() {
        mPresenter = new SettingsPresenter<>();
        mPresenter.attach(this);
    }

    @Override
    public void setListPreferenceSummary(Preference preference, Object newValue) {
        ListPreference listPreference = (ListPreference) preference;
        CharSequence[] entries = listPreference.getEntries();
        listPreference.setSummary(entries[listPreference.findIndexOfValue((String) newValue)]);
    }

    @Override
    public void setCheckBoxPreferenceSummary(Preference preference, Object newValue) {
        CheckBoxPreference checkBoxPreference = (CheckBoxPreference) preference;
        checkBoxPreference.setSummary(checkBoxPreference.getSummary());
    }

    @Override
    public void setSwitchPreferenceSummary(Preference preference, Object newValue) {
        SwitchPreference switchPreference = (SwitchPreference) preference;
        switchPreference.setSummary(switchPreference.getSummary());
    }

    @Override
    public void setupSimSelection() {
        ListPreference simSelectionPreference = findPreference(getString(R.string.pref_sim_select_key));

        @SuppressLint("MissingPermission")
        List<SubscriptionInfo> subscriptionInfoList = SubscriptionManager.from(getContext()).getActiveSubscriptionInfoList();
        int simCount = subscriptionInfoList.size();

        if (simCount == 1) {
            simSelectionPreference.setSummary(getString(R.string.pref_sim_select_disabled));
            simSelectionPreference.setEnabled(false);
        } else {
            List<CharSequence> simsEntries = new ArrayList<>();

            for (int i = 0; i < simCount; i++) {
                SubscriptionInfo si = subscriptionInfoList.get(i);
                Timber.i("Sim info " + i + " : " + si.getDisplayName());
                simsEntries.add(si.getDisplayName());
            }

            CharSequence[] simsEntriesList = simsEntries.toArray(new CharSequence[simsEntries.size()]);
            simSelectionPreference.setEntries(simsEntriesList);
            // simsEntries.add(getString(R.string.pref_sim_select_ask_entry));
            CharSequence[] simsEntryValues = {"0", "1"};
            simSelectionPreference.setEntryValues(simsEntryValues);
        }
    }

    @Override
    public void goToMainActivity() {
        Intent intent = new Intent(getActivity(), MainActivity.class);
        startActivity(intent);
    }


    @Override
    public boolean hasPermission(String permission) {
        return true;
    }

    @Override
    public boolean hasPermissions(String[] permissions) {
        return true;
    }

    @Override
    public void askForPermission(String permission, int requestCode) {
        requestPermissions(new String[]{permission}, requestCode);
    }

    @Override
    public void askForPermissions(String[] permissions, int requestCode) {
        requestPermissions(permissions, requestCode);
    }

    @Override
    public void showMessage(String message) {
        Toast.makeText(getContext(), message, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showMessage(int stringResId) {
        showMessage(getString(stringResId));
    }

    @Override
    public void showError(String message) {
        showMessage(message);
    }

    @Override
    public void showError(int stringResId) {
        showMessage(stringResId);
    }
}

File: app/src/main/java/com/chooloo/www/callmanager/ui/settings_fragment/SettingsFragment.kt
package com.chooloo.www.callmanager.ui.settings_fragment

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.telephony.SubscriptionManager
import android.view.View
import android.widget.Toast
import androidx.preference.*
import com.chooloo.www.callmanager.R
import com.chooloo.www.callmanager.ui.main.MainActivity
import timber.log.Timber
import java.util.*

class SettingsFragment : PreferenceFragmentCompat(), SettingsMvpView {
    private lateinit var _presenter: SettingsPresenter<SettingsMvpView>

    companion object {
        fun newInstance(): SettingsFragment = SettingsFragment()
    }

    override fun onCreatePreferences(savedInstanceState: Bundle, rootKey: String) {
        setPreferencesFromResource(R.xml.preference, rootKey)
        findPreference<Preference>(getString(R.string.pref_app_color_key))?.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { preference, newValue ->
            _presenter.refresh()
            _presenter.onListPreferenceChange(preference, newValue)
        }
        findPreference<Preference>(getString(R.string.pref_app_theme_key))?.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { preference, newValue ->
            _presenter.refresh()
            _presenter.onListPreferenceChange(preference, newValue)
        }
        findPreference<Preference>(getString(R.string.pref_reject_call_timer_key))?.setOnPreferenceChangeListener { preference, newValue -> _presenter.onListPreferenceChange(preference, newValue) }
        findPreference<Preference>(getString(R.string.pref_answer_call_timer_key))?.setOnPreferenceChangeListener { preference, newValue -> _presenter.onListPreferenceChange(preference, newValue) }
        findPreference<Preference>(getString(R.string.pref_default_page_key))?.setOnPreferenceChangeListener { preference, newValue -> _presenter.onListPreferenceChange(preference, newValue) }
        findPreference<Preference>(getString(R.string.pref_sim_select_key))?.setOnPreferenceChangeListener { preference, newValue -> _presenter.onListPreferenceChange(preference, newValue) }
        findPreference<Preference>(getString(R.string.pref_is_biometric_key))?.setOnPreferenceChangeListener { preference, newValue -> _presenter.onSwitchPreferenceChange(preference, newValue) }

        setupSimSelection()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        _presenter.onRequestPermissionResult(requestCode, grantResults)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        onSetup()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _presenter.detach()
    }

    override fun onSetup() {
        _presenter = SettingsPresenter()
        _presenter.attach(this)
    }

    override fun setListPreferenceSummary(preference: Preference, newValue: Any) {
        val listPreference = preference as ListPreference
        val entries = listPreference.entries
        listPreference.summary = entries[listPreference.findIndexOfValue(newValue as String)]
    }

    override fun setCheckBoxPreferenceSummary(preferenece: Preference, newValue: Any) {
        val checkBoxPreference = preferenece as CheckBoxPreference
        checkBoxPreference.summary = checkBoxPreference.summary
    }

    override fun setSwitchPreferenceSummary(preference: Preference, newValue: Any) {
        val switchPreference = preference as SwitchPreference
        switchPreference.summary = switchPreference.summary
    }

    override fun setupSimSelection() {
        TODO("implement normally")
        val simSelectionPreference = findPreference<ListPreference>(getString(R.string.pref_sim_select_key))

        @SuppressLint("MissingPermission")
        val subscriptionInfoList = activity?.getSystemService(SubscriptionManager::class.java)?.activeSubscriptionInfoList
        val simCount = subscriptionInfoList?.size

        if (simCount == 1) {
            simSelectionPreference?.summary = getString(R.string.pref_sim_select_disabled)
            simSelectionPreference?.isEnabled = false
        } else if (simCount != null) {
            val simsEntries: MutableList<CharSequence> = ArrayList()
            for (i in 0 until simCount) {
                val si = subscriptionInfoList[i]
                Timber.i("Sim info " + i + " : " + si.displayName)
                simsEntries.add(si.displayName)
            }
            val simsEntriesList = simsEntries.toTypedArray()
            simSelectionPreference?.entries = simsEntriesList
            // simsEntries.add(getString(R.string.pref_sim_select_ask_entry));
            val simsEntryValues = arrayOf<CharSequence>("0", "1")
            simSelectionPreference?.entryValues = simsEntryValues
        }
    }

    override fun goToMainActivity() {
        startActivity(Intent(activity, MainActivity::class.java))
    }

    override fun hasPermission(permission: String): Boolean {
        return true
    }

    override fun hasPermissions(permissions: Array<String>): Boolean {
        return true
    }

    override fun askForPermission(permission: String, requestCode: Int) {
        requestPermissions(arrayOf(permission), requestCode)
    }

    override fun askForPermissions(permissions: Array<String>, requestCode: Int) {
        requestPermissions(permissions, requestCode)
    }

    override fun showMessage(message: String) {
        Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
    }

    override fun showMessage(stringResId: Int) {
        showMessage(getString(stringResId))
    }

    override fun showError(message: String) {
        showMessage(message)
    }

    override fun showError(stringResId: Int) {
        showMessage(stringResId)
    }
}